lightningd/channel/channel.c

```diff
@@ -523,7 +527,89 @@ static void our_htlc_failed(struct peer *peer, const struct htlc *htlc)
 
 static void their_htlc_locked(struct peer *peer, const struct htlc *htlc)
 {
-	status_trace("FIXME: their htlc %"PRIu64" locked", htlc->id);
+	tal_t *tmpctx = tal_tmpctx(peer);
+	u8 *msg;
+	struct onionpacket *op;
+	struct route_step *rs;
+	struct sha256 ss, bad_onion_sha;
+	enum onion_type failcode;
+	enum channel_remove_err rerr;
+	struct pubkey ephemeral;
+
+	status_trace("their htlc %"PRIu64" locked", htlc->id);
+
+	/* We unwrap the onion now. */
+	/* FIXME: We could do this earlier and call HSM async, for speed. */
+	op = parse_onionpacket(tmpctx, htlc->routing, TOTAL_PACKET_SIZE);
+	if (!op) {
+		/* FIXME: could be bad version, bad key. */
+		failcode = WIRE_INVALID_ONION_VERSION;
+		goto bad_onion;
+	}
+
+	/* Because wire takes struct pubkey. */
+	ephemeral.pubkey = op->ephemeralkey;
+	msg = towire_hsm_ecdh_req(tmpctx, &ephemeral);
+	if (!wire_sync_write(HSM_FD, msg))
+		status_failed(WIRE_CHANNEL_HSM_FAILED, "Writing ecdh req");
+	msg = wire_sync_read(tmpctx, HSM_FD);
+	if (!msg || !fromwire_hsm_ecdh_resp(msg, NULL, &ss))
+		status_failed(WIRE_CHANNEL_HSM_FAILED, "Reading ecdh response");
+
+	if (memeqzero(&ss, sizeof(ss))) {
+		failcode = WIRE_INVALID_ONION_KEY;
+		goto bad_onion;
+	}
+
+	rs = process_onionpacket(tmpctx, op, ss.u.u8, htlc->rhash.u.u8,
+				 sizeof(htlc->rhash));
+	if (!rs) {
+		failcode = WIRE_INVALID_ONION_HMAC;
+		goto bad_onion;
+	}
+
+	/* Unknown realm isn't a bad onion, it's a normal failure. */
+	/* FIXME: Push complete hoppayload up and have master parse? */
```

Comment:
Pushing up allows us to create simple bridges between networks by having `lightningd` call out to an external handler :+1: